Bypass SchemaForm

cvalarida opened the pull request:
## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/27148

This PR adds functionality to bypass the `SchemaForm` in the schema-based forms library. See the original issue for details.

## Testing done
- Lots of unit tests
- I created a new form to test this with and poked around with that

Commit:
Render the custom page instead of SchemaForm

Commit:
Pass the right props to the CustomPage

Commit:
Pass the CustomPage to the helper functions

Commit:
Optionally chain the formContext in case it's undefined

Commit:
Return all the entire form data to custom page when showPagePerItem is true

Commit:
Refactor: Get page contents in its own function

Commit:
Refactor: Pull out page content into its own function

Commit:
Call getCustomPageContent() for custom pages

Commit:
Test: Throw error when CustomPageReview doesn't exist

Commit:
Refactor test

Commit:
onSubmit goes to the next page, not callOnContinue

Commit:
Expose CustomPage and CustomPageReview on routes

Commit:
A little better error message

Commit:
Refactor to render either custom components

Commit:
Use this.getCustomPageContent()

No sense re-inventing the wheel.

Commit:
Add comment

Commit:
Add props to render the correct buttons

Commit:
Add formConfig validation during createRoutes

Commit:
refactor: noVisibleFields -> visibleFields

If `pageSchema` is `undefined` as it is when schema.properties is empty,
`noVisibleFields` evaluates to false. The double negative meant that
there were visible fields (allegedly), so it was trying to render the
SchemaForm whithout any properties.

Commit:
Pull out the form nav buttons for use in CustomPages

Commit:
Rename onSubmit to updatePage

Commit:
Update JSDoc for getPages

cvalarida commented on src/platform/forms-system/src/js/review/ReviewCollapsibleChapter.jsx:
```diff
@@ -83,208 +83,253 @@ class ReviewCollapsibleChapter extends React.Component {
 
     const classes = classNames('usa-accordion-bordered', 'form-review-panel', {
-      'schemaform-review-chapter-warning': hasUnviewedPages,
+      'schemaform-review-chapter-warning': this.props.hasUnviewedPages,
```
No change here: I'm just not deconstructing `this.props` to get it.

cvalarida commented on src/platform/forms-system/src/js/review/ReviewCollapsibleChapter.jsx:
```diff
@@ -83,208 +83,253 @@ class ReviewCollapsibleChapter extends React.Component {
+      ? title(pageData)
+      : title) || 'page'}`;
+
+    const visibleFields =
+      pageSchema &&
+      Object.entries(pageSchema.properties).filter(([propName]) =>
+        showReviewField(
+          propName,
+          pageSchema,
+          pageUiSchema,
+          form.data,
+          formContext,
+        ),
+      ).length > 0;
```
This changed from the original. Previously, it was `noVisibleFields`. I changed to `visibleFields` because:
- It's easier to reason about.
- When `pageSchema` is `undefined`, such as when we've got an empty `schema.properties` because we have a `CustomPage`, the previous statement (`pageSchema && !Object.entries(...).filter(...).length > 0`) was evaluating to false—so `!noVisibleFields` because `pageSchema` doesn't exist. I needed it to mean "there are no visible fields when there's no schema," so I refactored and fixed that little bug.

cvalarida commented on src/platform/forms-system/src/js/review/ReviewCollapsibleChapter.jsx:
```diff
@@ -83,208 +83,253 @@ class ReviewCollapsibleChapter extends React.Component {
+    return chapterTitle;
+  };
 
+  getSchemaformPageContent = (page, props) => {
```
This is refactored from what was in `render()` with a few changes called out in the review comments.

cvalarida commented on src/platform/forms-system/src/js/review/ReviewCollapsibleChapter.jsx:
```diff
@@ -83,208 +83,253 @@ class ReviewCollapsibleChapter extends React.Component {
+        this.getCustomPageContent(page, props)
+      );
+
+    const renderCustom = editing ? !!page.CustomPage : !!page.CustomPageReview;
+    const reviewContent = renderCustom ? customContent() : schemaformContent();
```
This is new. This function (`getSchemaformPageContent`) is called when there is no `CustomPageReview`, so when in "view" mode, it renders the `SchemaForm` in review mode.

These two lines here determine what to render in the edit mode.

cvalarida reviewed (commented):
This PR is a lot of refactoring, some new code, a lot of testing, and a lot of thinking.

The new features introduced in this PR:
- We can supply a `CustomPage` React component to bypass the `SchemaForm`
- We can leverage the existing review page "view" mode by optionally supplying the `schema` (and `uiSchema`) in addition to the `CustomPage`
- We can bypass the review page "view" mode by supplying a `CustomPageReview` React component

bkjohnson commented on src/platform/forms-system/src/js/containers/FormPage.jsx:
```diff
@@ -140,6 +146,30 @@ class FormPage extends React.Component {
+    // Bypass the SchemaForm and render the custom component
+    // NOTE: I don't think FormPage is rendered on the review page, so I believe
+    // onReviewPage will always be false here
+    if (typeof route.pageConfig.CustomPage === 'function') {
```
I think we have a suitable helper for this that catches some edge cases:

https://github.com/department-of-veterans-affairs/vets-website/blob/4f4d329f489626cc650c3a1935616340c5ad5daa/src/platform/utilities/ui/index.js#L89-L96

cvalarida replied:
Oh hey! Nice find! That's much nicer.

bkjohnson commented on src/platform/forms-system/src/js/containers/FormPage.jsx:
```diff
@@ -140,6 +146,30 @@ class FormPage extends React.Component {
+      return (
+        <div className={pageClasses}>
+          <route.pageConfig.CustomPage
+            name={route.pageConfig.pageKey}
+            title={route.pageConfig.title}
+            data={data}
+            pagePerItemIndex={params ? params.index : undefined}
+            isLoggedIn={formContext?.isLoggedIn}
+            prefilled={formContext?.prefilled}
+            onReviewPage={formContext?.onReviewPage}
+            trackingPrefix={this.props.form.trackingPrefix}
+            uploadFile={this.props.uploadFile}
+            onChange={this.onChange}
+            goBack={this.goBack}
+            goForward={this.onSubmit}
```
Where is this API coming from?

cvalarida replied:
My head. :stuck_out_tongue_closed_eyes: 

It's really whatever we decide it is. I put these in here because I thought they'd be the most useful. The way I came up with these is by copying the props to `SchemaForm` and going through the list, determining which ones would be useful. I renamed `onSubmit` and didn't include some of the others like `callOnContinue` which just served as a way to call some functions defined in the `formConfig`.

bkjohnson commented on src/platform/forms-system/src/js/containers/FormPage.jsx:
```diff
@@ -140,6 +146,30 @@ class FormPage extends React.Component {
+    // onReviewPage will always be false here
+    if (typeof route.pageConfig.CustomPage === 'function') {
+      return (
+        <div className={pageClasses}>
```
Does the wrapping div do anything for us here? Could the classes be applied directly to `CustomPage`?

cvalarida replied:
It doesn't get us _much,_ no. I don't know what happens if all pages aren't wrapped in a `form-panel` class, so I opted to keep it because it didn't seem like it'd hurt anything. :man_shrugging:

bkjohnson reviewed (commented):
I'm going to need to take a closer look at the changes to `ReviewCollapsibleChapter`

cvalarida commented on src/platform/forms-system/src/js/validate-config/README.md:
```diff
@@ -0,0 +1,2 @@
+# Form Config Validation
+Each form config validation function must accept the `formConfig` and throw an error if one is found, or return `undefined`.
```
No. Clarification: "...must accept the `formConfig` and throw an error if an error in the config is found, or return `undefined` if there is no error."

cvalarida commented on src/platform/forms-system/src/js/validate-config/get-pages.js:
```diff
@@ -0,0 +1,30 @@
+/**
+ * A helper function to easily iterate over each page in a form config and
+ * preserve the path for debugging output.
```
Yeah, I need easy access to the configs. That's the whole point of `getPages()`. :man_shrugging:

cvalarida commented on src/platform/forms-system/src/js/validate-config/index.js:
```diff
@@ -0,0 +1,7 @@
+import validateCustomPages from './validate-custom-pages';
+
+const validateConfig = formConfig => {
+  validateCustomPages(formConfig);
+};
```
As needed, but probably not from this PR. This is just a convenient place to add more config validation functions later.

cvalarida commented on src/platform/forms-system/test/js/review/ReviewCollapsibleChapter.unit.spec.jsx:
```diff
@@ -814,4 +814,48 @@ describe('<ReviewCollapsibleChapter>', () => {
+        ),
+      ).to.throw('CustomPage found for test, but missing CustomPageReview.');
+    });
+    it('should render CustomPageReview when CustomPage is present', () => {});
+    it('should render a CustomPageReview for each item in an array when showPagePerItem is true', () => {});
```
Heh, this is me keeping track of what I still need to test for. I'm writing these now. :sweat_smile:

Commit:
Use react testing library

Because two rendering frameworks aren't enough!

Commit:
Update test with more state

Because we're no longer rendering anything if there's no schema found,
we have to update the test to give it schema.

Commit:
Update test to use FormNavButtons

Commit:
Use FormNavButtons in this test too...

Because we're testing implementation details. :shrug:

Commit:
Refactor: Handle edit state in getChapterContent

Commit:
Add test: Render CustomPage in edit mode

Commit:
Remove dead code

CustomPageReview will not be null when this is called

Commit:
Add more props to the CustomPage from the review

Commit:
FormNavButtons documentation

Commit:
Validate Config README

Commit:
Use isReactComponent

Commit:
Don't setData for CustomPages in showPagePerItem

Commit:
Just toggle the edit state on updatePage

`this.handleSubmit` was also setting the data to outdated information,
so...stop that.

Commit:
Fill in the CustomPageReview props

Commit:
Render a new CustomPageReview for each item in an array

Commit:
Add schema & uiSchema to test pageConfig

Commit:
Render nothing test

Commit:
Look for children length of 0 instead

Commit:
Render

Commit:
remove comment

Commit:
Add edit button test

Commit:
Remove outdated comment

Commit:
Page update function test

Commit:
Pass form data test stubs

Commit:
Custom page review data test

Commit:
CustomPage data test

Commit:
Add the editPage function to CustomPageReview

Commit:
Match CustomPage props

`ReviewCollapsibleChapter` and `FormPage` were sending different props
to `CustomPage`. They now match. The `formContext` props can be gathered
by just connecting the `CustomPage` to the redux store, and the
`onChange` can be achieved by directly calling `setData`.

bkjohnson commented on src/platform/forms-system/src/js/containers/FormPage.jsx:
```diff
@@ -157,27 +187,11 @@ class FormPage extends React.Component {
+          <FormNavButtons
+            goBack={!isFirstRoutePage && this.goBack}
+            goForward={callOnContinue}
+            submitToContinue
```
Is this prop really necessary? Looks like having `submitButton` hardcoded to `true` in `FormNavButtons` would eliminate the need for this.

cvalarida replied:
True, but that's not necessarily _how_ `CustomPage`s will use the navigation buttons. If, for example, the page doesn't have a `<form>` element, but just contains some text, having a `submit` button to continue to the next page won't work because the form submit action won't navigate. Instead, the `CustomPage` would have to render the continue button as a `<button type="button" onClick={goForward}>` (more or less) to navigate to the next page.

bkjohnson commented on src/platform/forms-system/src/js/containers/FormPage.jsx:
```diff
@@ -121,7 +130,7 @@ class FormPage extends React.Component {
     const data = this.formData();
 
-    if (route.pageConfig.showPagePerItem) {
+    if (route.pageConfig.showPagePerItem && !route.pageConfig.CustomPage) {
```
Is this just for performance reasons? Looks like the only things happening in the `if` block are assigning values to `schema` & `uiSchema`, which as I understand it aren't used in a `CustomPage`.

_Rereads code before submitting review_

Oh, maybe it's because we're not expecting `route.pageConfig.arrayPath` to exist if there is a `CustomPage`?

cvalarida replied:
The array path may exist, but the schema properties may be empty, so trying to get the `.items` will result in an error (unless we optionally chain, of course). And yeah, we don't use it anyhow. :man_shrugging:

bkjohnson commented on src/platform/forms-system/src/js/review/ReviewCollapsibleChapter.jsx:
```diff
@@ -83,208 +83,257 @@ class ReviewCollapsibleChapter extends React.Component {
+            ? pageConfig.editMode[page.index]
+            : pageConfig.editMode;
+
+          if (editing) {
+            return pageConfig.CustomPage
+              ? this.getCustomPageContent(page, props, editing)
+              : this.getSchemaformPageContent(page, props, editing);
+          } else {
+            return pageConfig.CustomPageReview
+              ? this.getCustomPageContent(page, props, editing)
+              : this.getSchemaformPageContent(page, props, editing);
+          }
```
This is unfortunate. It looks like there are only two possible outcomes, but there are 4 paths (two for each)

What do you think about this (assuming I've done this right)?


```js

const showCustomPage = 
    (editing && pageConfig.CustomPage)
    || (!editing && pageConfig.CustomPageReview);

return showCustomPage 
    ? this.getCustomPageContent(page, props, editing)
    : this.getSchemaformPageContent(page, props, editing)
```

cvalarida replied:
Actually, reading yours, I think we can come up with a little better:
```js
const showCustomPage = editing
    ? !!pageConfig.CustomPage
    : !!pageConfig.CustomPageReview;

return showCustomPage 
    ? this.getCustomPageContent(page, props, editing)
    : this.getSchemaformPageContent(page, props, editing);
```

bkjohnson commented on src/platform/forms-system/test/js/review/ReviewCollapsibleChapter.unit.spec.jsx:
```diff
@@ -814,4 +822,246 @@ describe('<ReviewCollapsibleChapter>', () => {
+        pageKey: 'test',
+        CustomPage: pages[0].CustomPage,
+        CustomPageReview: pages[0].CustomPageReview,
+        index: 1,
```
It seems strange to me that we have an array of objects, where one of the keys in each object is `index`. But I think that's a larger forms library problem.

cvalarida replied:
Yeah, I agree. It has to do with the way the forms library deconstructs the `formConfig` and shunts different pieces of it around in different forms. :man_shrugging: Apparently, somewhere along the way, the `index` gets added to this. You can see where `page.index` is used in `ReviewCollapsibleChapter` for confirmation that it's a Thing™. I didn't make it up. :man_shrugging:

bkjohnson commented on src/platform/forms-system/test/js/review/ReviewCollapsibleChapter.unit.spec.jsx:
```diff
@@ -170,17 +172,23 @@ describe('<ReviewCollapsibleChapter>', () => {
         showPagePerItem: true,
         arrayPath: 'testing',
         path: 'path/:index',
+        index: 1,
```
I know that `page.index` existed before this PR, but why do we need these now? Would existing configs be considered broken without this?

cvalarida replied:
Yes.

cvalarida replied:
To be clear, though, this isn't specified in the `formConfig`s. These come from the way the forms library hydrates array routes. I'm unclear on the details, but essentially it duplicates each page config with its own `index` for each item in the array at `arrayPath`.

bkjohnson reviewed (commented):
Got a few questions which might reveal me being a forms noob :sweat_smile:

Commit:
Merge branch 'master' into bypass-schemaform-cv

Commit:
Refactor branching logic to either custom or schemaform

Commit:
Merge branch 'master' into bypass-schemaform-cv

bkjohnson reviewed (approved):
![Take this show on the road](https://media.giphy.com/media/3oKIPhPEOHJLJoVT3i/giphy.gif)
